src/coreclr/jit/importer.cpp

```diff
@@ -1094,47 +1099,50 @@ GenTree* Compiler::impAssignStructPtr(GenTree* destAddr, GenTree* src, unsigned
 }
 
 //------------------------------------------------------------------------
-// impGetStructAddr: Get the address of a struct value.
+// impGetNodeAddr: Get the address of a value.
 //
 // Arguments:
-//    structVal - The value in question
-//    curLevel  - Stack level for spilling
-//    willDeref - Whether the caller will dereference the address
+//    val         - The value in question
+//    curLevel    - Stack level for spilling
+//    pDerefFlags - Flags to be used on dereference, nullptr when
+//                 the address won't be dereferenced
 //
 // Return Value:
-//    In case "structVal" can represent locations (is an indirection/local),
+//    In case "val" can represent locations (is an indirection/local),
 //    will return its address. Otherwise, address of a temporary assigned
-//    the value of "structVal" will be returned.
+//    the value of "val" will be returned.
 //
-GenTree* Compiler::impGetStructAddr(GenTree* structVal, unsigned curLevel, bool willDeref)
+GenTree* Compiler::impGetNodeAddr(GenTree* val, unsigned curLevel, GenTreeFlags* pDerefFlags)
 {
-    assert(varTypeIsStruct(structVal));
-    switch (structVal->OperGet())
+    switch (val->OperGet())
     {
         case GT_BLK:
         case GT_IND:
-            if (willDeref)
+            if (pDerefFlags != nullptr)
             {
-                return structVal->AsIndir()->Addr();
+                *pDerefFlags |= (val->gtFlags & GTF_IND_FLAGS);
+                return val->AsIndir()->Addr();
```

Comment:
I think we should limit the set of flags here to `(GTF_IND_VOLATILE | GTF_IND_NONFAULTING | GTF_IND_UNALIGNED | GTF_IND_INITCLASS)`, same as physical promotion does. The other flags are more likely to be incorrect w.r.t. applying them to arbitrary callers. This set should also be documented in the method header.

Also, the Jit convention is that we should initialize `pDerefFlags` to `GTF_EMPTY` if passed (s. t. the callers don't have to).